nautobot/extras/groups.py

```diff
@@ -0,0 +1,182 @@
+import logging
+import urllib
+
+from django import forms
+from django.db import models
+from django.urls import reverse
+from django.utils.functional import classproperty
+
+from nautobot.utilities.utils import get_filterset_for_model, get_form_for_model, get_route_for_model
+
+
+logger = logging.getLogger(__name__)
+
+
+def dynamicgroup_map_factory(model):
+    """Generate a `FooDynamicGroupMap` class for a given `model`."""
+
+    try:
+        if not issubclass(model, models.Model):
+            return None
+    except TypeError:
+        return None
+
+    filterset_class = get_filterset_for_model(model)
+    filterform_class = get_form_for_model(model, form_prefix="Filter")
+    form_class = get_form_for_model(model)
+
+    group_map = type(
+        str("%sDynamicGroupMap" % model._meta.object_name),
+        (BaseDynamicGroupMap,),
+        {
+            "model": model,
+            "filterset_class": filterset_class,
+            "filterform_class": filterform_class,
+            "form_class": form_class,
+        },
+    )
+
+    return group_map
+
+
+class BaseDynamicGroupMap:
+    """
+    Dynamic Group mapping used to generate mappings for each model class.
+
+    This class itself should not be invoked directly as the class variables will
+    not be populated and most class methods will fail.
+    """
+
+    model = None
+    form_class = None
+    filterset_class = None
+    filterform = None
+
+    # This is used as a `startswith` check on field names, so these can be
+    # explicit fields or just substrings.
+    #
+    # Currently this means skipping "search", custom fields, and custom relationships.
+    #
+    # FIXME(jathan): As one example, `DeviceFilterSet.q` filter searches in `comments`. The issue
+    # really being that this field renders as a textarea and it's not cute in the UI. Might be able
+    # to dynamically change the widget if we decide we do want to support this field.
+    #
+    # Type: tuple
+    exclude_filter_fields = ("q", "cr", "cf", "comments")  # Must be a tuple
+
+    @classproperty
+    def base_url(cls):
+        if cls.model is None:
+            return
+        route_name = get_route_for_model(cls.model, "list")
+        return reverse(route_name)
+
+    @classmethod
+    def fields(cls):
+        """Return all FilterForm fields in a dictionary."""
+
+        # Get model form and fields
+        modelform = cls.form_class()
+        modelform_fields = modelform.fields
+
+        # Get filter form and fields
+        filterform = cls.filterform_class()
+        filter_fields = filterform.fields
+
+        # Get filterset and fields
+        filterset = cls.filterset_class()
+        filterset_fields = filterset.filters
```

Comment:
Yeah this is weird. Because there's no use-case for using a bare `BaseDynamicGroupMap`. It is expected that they will be provided by `dynamicgorup_map_factory()`. This entire pattern is relying on this just being a class full of methods/properties without any instance-level required arguments for validation of correctness.

This actually just further gels the fact that `BaseDynamicGroupMap` should go away entirely and, and all of the `DynamicGroup.map` calls just converted to internal calls on a `DynamicGroup` model instance.